prov/util/src/util_cuda_mem_monitor.c

```diff
@@ -0,0 +1,416 @@
+/*
+ * (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_mr.h"
+
+#ifdef HAVE_LIBCUDA
+
+#include "ofi_tree.h"
+#include "ofi_iov.h"
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+struct ofi_cuda_mm_entry {
+	struct iovec iov;
+	uint64_t id;
+	struct dlist_entry lru_entry;
+	struct ofi_rbnode *node;
+};
```

Comment:
This looks like it's duplicating the functionality of the mr cache.  Specifically, this is similar to ofi_mr_entry.  Only the cache should need to maintain a rbtree and lru list.  We shouldn't require 2 searches for the same address range.  If hmem requires storing some additional amount of data per ofi_mr_entry (such as an id), let's put that data there.

The ofi_mr_entry is currently broken into 2 pieces.  There's the portion used by the cache itself, and a secondary data section for use by the provider.  It seems that we just need a third section that's HMEM specific.  Although, we could allocate this dynamically, it doesn't seem worth it at this point.  For example:

```
union ofi_mr_hmem_info {
    uint64_t cuda_id;
};

struct ofi_mr_entry {
    struct ofi_mr_info info;
    void  *storage_context;
    unsigned int subscribed:1;
    int use_cnt;
    struct dlist_entry list_entry;
    enum fi_hmem_iface iface;
    union ofi_mr_hmem_info hmem_info;
    uint8_t data[];
};
```

This moves the iface from ofi_mr_info here.  I haven't looked close enough at the patches to see if ofi_mr_info really needs this data, rather than having it here.  This requires changing the monitor valid() handler to accept the hmem_info as input.  Monitors that do not use this data can ignore it.

Reply:
Sounds good.